Implemented ReflectiveShell as optional boundary

ehlertdo opened the pull request:
Hi,
I needed a reflective boundary in my 3D simulations but in the form of a sphere instead of the currently available ReflectiveBox. Maybe this feature could be of more general interest since astrophysical environments are often more spherical than they are boxy.
Note that the shell reflects in both directions, i.e. outside particles are kept out and inside particles are kept in.
Best,
Domenik

PS: I copied the distance() and normal() functions from the Sphere class in Geometry.cpp but probably there's a more elegant way to do this with inheritance.

![cr_tracks](https://github.com/user-attachments/assets/ff936d0d-0a70-4636-94f0-782a36c8b337)
![cr_tracks](https://github.com/user-attachments/assets/1b29f7f1-c217-4551-96b0-b06f3eed7705)

JulienDoerner reviewed (changes requested):
Hey @ehlertdo, 

Thanks for sharing your implementation. In general, the tool looks good for me. 

I have tested it and found some deviation when using large steps. I guess this is caused by the position where you calculate the normal. This is given at the current position which is not necessarily a position on the sphere. Would you have an idea how to fix this, or can this be neglected in regular cases?
![image](https://github.com/user-attachments/assets/4a00bf55-91d1-4533-9ddf-78b07215ba8e)

Beside this I would ask you to add your changes to the Changes.md and maybe you can provide a simple test case as a notebook and as unit test.

ehlertdo commented:
Hi Julien,
I did indeed approximate the true normal to the sphere at the point of intersection by the normal at the final point of the particle. I have now added a few lines to calculate the intersection point before constructing the normal which, in my tests reduces the error at larger step sizes significantly. I also added a unit test for the ReflectiveShell and updated the changelog and doc page.

JulienDoerner commented on include/crpropa/module/Boundary.h:
```diff
@@ -37,6 +37,32 @@ class PeriodicBox: public Module {
+	 @param cen	vector corresponding to the center of the sphere
+	 @param r	value corresponding to the radius of the shell
+	 */
+	ReflectiveShell(Vector3d cen, double r);
```
Code Style: use full words for variable names: `center` instead of `cen`

JulienDoerner commented on src/module/Boundary.cpp:
```diff
@@ -40,6 +40,58 @@ std::string PeriodicBox::getDescription() const {
 }
 
+
+ReflectiveShell::ReflectiveShell(Vector3d cen, double r) :
```
Code Style

JulienDoerner commented on src/module/Boundary.cpp:
```diff
@@ -40,6 +40,58 @@ std::string PeriodicBox::getDescription() const {
+		Vector3d previousPosition = c->previous.getPosition();
+		// get point where trajectory intersects the shell boundary
+		double p_half = previousPosition.dot(currentDirection) / currentDirection.getR2();
+		double q = (previousPosition.getR2() - radius*radius) / currentDirection.getR2();
```
Code Style: spaces around operators. Use `a * b` instead of `a*b`. 
This also happens below. I mark it only once here.

JulienDoerner commented on src/module/Boundary.cpp:
```diff
@@ -40,6 +40,58 @@ std::string PeriodicBox::getDescription() const {
+std::string ReflectiveShell::getDescription() const {
+	std::stringstream ss;
+	ss << "Reflective sphere: " << std::endl
+			<< "   Center: " << center << std::endl
```
Is it useful to define some default unit for the printing like kpc or Mpc?

JulienDoerner commented on test/testBreakCondition.cpp:
```diff
@@ -345,6 +345,31 @@ TEST(PeriodicBox, low) {
+
+	shell.process(&c);
+
+	EXPECT_NEAR(89.9965, c.current.getPosition().x, 1e-4);
```
Please leave a comment in the code on how you get to those expected values (or even calculate them with the trigonometric functions here).
